File: geral/livro_entendendo_algoritmos/linked_list.rb
class Node
    attr_accessor :value, :after

    def initialize(value, after)
        @value = value
        @after = after
    end
end

class LinkedList
    attr_accessor :head

    def prepend(value)
        @head = Node.new(value, @head)
    end

    def append(value)
        if @head == nil
            @head = Node.new(value, nil)
        end

        current = @head

        while current.after != nil
            current = current.after
        end

        current.after = Node.new(value, nil)
    end

    def remove(value)
        if @head == nil
            return
        end

        if @head.value = value
            @head = @head.after
        end

        current = @head

        while current.after != nil
            if current.after.value == value
                current.after = current.after.after
                return
            end

            current = current.after
        end
    end
end

def main()
    list = LinkedList.new

    list.prepend 10

    list.prepend 20

    list.prepend 30

    list.append 40

    list.append 50

    list.remove 20

    list.remove 30

    list.remove 10

    print list
end

def print(list)
    current = list.head

    while current != nil
        puts current.value

        current = current.after
    end
end

main

File: geral/livro_entendendo_algoritmos/linkedlist/linkedlist.go
package main

import "fmt"

type Node struct {
	Value int
	Next  *Node
}

func NewNode(value int, next *Node) *Node {
	return &Node{
		Value: value,
		Next:  next,
	}
}

type LinkedList struct {
	Head *Node
}

func NewLinkedList() *LinkedList {
	return &LinkedList{}
}

func (l *LinkedList) Prepend(value int) {
	l.Head = NewNode(value, l.Head)
}

func (l *LinkedList) Append(value int) {
	if l.Head == nil {
		l.Head = NewNode(value, nil)
		return
	}

	current := l.Head

	for current.Next != nil {
		current = current.Next
	}

	current.Next = NewNode(value, nil)
}

func (l *LinkedList) Remove(value int) {
	if l.Head == nil {
		return
	}

	if l.Head.Value == value {
		l.Head = l.Head.Next
		return
	}

	current := l.Head

	for current.Next != nil {
		if current.Next.Value == value {
			current.Next = current.Next.Next
			return
		}

		current = current.Next
	}
}

func main() {
	list := NewLinkedList()

	list.Prepend(10)

	list.Prepend(20)

	list.Prepend(30)

	list.Append(50)

	list.Append(60)

	list.Remove(20)

	list.Remove(10)

	list.Remove(30)

	print(list)
}

func print(list *LinkedList) {
	current := list.Head

	for current != nil {
		fmt.Printf("%d\n", current.Value)

		current = current.Next
	}
}
